### PR TITLE
Switch to `pyproject.toml`

### PR DESCRIPTION
**Pull request by MetRonnie**
Move from `setup.py` to `pyproject.toml`

Docs:
- https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html
- https://packaging.python.org/en/latest/specifications/declaring-project-metadata/

Also:
- Support Python 3.12
- GH Actions improvements

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -38,27 +48,48 @@ jobs:
...
+        run: pytest
 
-      - name: Upload coverage report
+      - name: Coverage report
         if: matrix.coverage
         run: |
-          coverage xml --ignore-errors
-          bash <(curl -s https://codecov.io/bash)
+          coverage xml
+          coverage report
```
Comment (oliver-sanders) [inline .github/workflows/test.yml:66]: Do we need this bit?

**Review by oliver-sanders — APPROVED** [review]

**Comment by oliver-sanders** [comment]
It might be worth taking the opportunity to roll the pytest.ini and .bandit files into this.

File: .github/workflows/test.yml (hunk 1 above)
Comment (MetRonnie) [reply .github/workflows/test.yml:66]: `coverage xml` - I think this is needed to convert the data into a compatible format for Codecov? But not sure.

`coverage report` - Not strictly needed but can be useful to see the coverage stats sometimes

**Comment by MetRonnie** [comment]
> (& .coveragerc maybe?).

Decided against this as coverage configs can be quite lengthy and also the exclusion regexes would need extra escaping in TOML

**Review by wxtim — APPROVED** [review]
- [x] Looks good
- [x] Tried it locally with a couple of different pythons

**Pull request merged**